### PR TITLE
Refactoring the fetching of products

### PR DESCRIPTION
**Pull request by loevgaard**
This PR will separate the population of products into multiple services

### CODE REVIEW

File: src/IndexResolver/ProductIndexesResolver.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
+<?php
```
Comment (loevgaard) [inline src/IndexResolver/ProductIndexesResolver.php:1]: Maybe this should be renamed to `ProductIndexScopesProvider`?

File: src/Message/Command/CommandInterface.php [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Setono\SyliusAlgoliaPlugin\Message\Command;
+
+interface CommandInterface
```
Comment (loevgaard) [inline src/Message/Command/CommandInterface.php:7]: The reason for adding this interface is that it makes it easy for plugin users to route all commands to an async transport

File: src/Repository/ProductRepositoryTrait.php [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+use Sylius\Bundle\ResourceBundle\Doctrine\ORM\EntityRepository;
+
+/**
+ * @mixin EntityRepository
+ */
+trait ProductRepositoryTrait
+{
+    public function createQueryBuilderFromProductIndexScope(ProductIndexScope $productIndexScope): QueryBuilder
+    {
+        // todo recheck this query
```
Comment (loevgaard) [inline src/Repository/ProductRepositoryTrait.php:19]: I am not the biggest Doctrine/SQL expert so I am not 100% sure this query is what we intend it to be

File: src/IndexResolver/ProductIndexesResolver.php (hunk 1 above)
Comment (Roshyo) [reply src/IndexResolver/ProductIndexesResolver.php:1]: Yes, it makes sense

File: src/Message/Command/CommandInterface.php (hunk 2 above)
Comment (Roshyo) [reply src/Message/Command/CommandInterface.php:7]: Ok, small amount of code, but still too early for this imo.

But let's keep it

File: src/Repository/ProductRepositoryTrait.php (hunk 3 above)
Comment (Roshyo) [reply src/Repository/ProductRepositoryTrait.php:19]: I would just avoid chaining calls, as it is not the best approach. But other than that, it seems to be correct

Comment (loevgaard) [reply src/Repository/ProductRepositoryTrait.php:19]: All I chain are query builder methods. Everything else I don't :D

Comment (loevgaard) [reply src/Repository/ProductRepositoryTrait.php:19]: Here's an article on it: https://ocramius.github.io/blog/fluent-interfaces-are-evil/

Comment (Roshyo) [reply src/Repository/ProductRepositoryTrait.php:19]: Of course he will say fluent is good on query builder... He is biased on this :D 

IMO it is not a good approach as any change in the middle that introduce bigger logic will result in git changelog being messed up...

Same for trailing coma in the end of PHP array, or now in method arguments

File: src/Provider/ProductIndexScopeProviderInterface.php [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+namespace Setono\SyliusAlgoliaPlugin\Provider;
+
+use Setono\SyliusAlgoliaPlugin\DTO\ProductIndexScope;
+
+interface ProductIndexScopeProviderInterface
+{
+    /**
+     * @return iterable<ProductIndexScope>
+     */
+    public function resolve(): iterable;
```
Comment (loevgaard) [inline src/Provider/ProductIndexScopeProviderInterface.php:14]: This method name should be renamed also to `getProductIndexScopes` or something like that

File: src/Repository/ProductRepositoryTrait.php (hunk 3 above)
Comment (loevgaard) [reply src/Repository/ProductRepositoryTrait.php:19]: :D

**Pull request merged**